README.md
# nuxt-font-loader-strategy

## :warning: **This project is no longer maintained**, because the concept is not suitable for global management of many fonts in larger projects. For this reason we have developed a new concept that guarantees smart, efficient and performant component-based font management even in larger websites. Please visit: https://github.com/GrabarzUndPartner/nuxt-speedkit

[![Grabarz & Partner - Module][grabarz-partner-module-src]][grabarz-partner-href] 

[![Main][github-workflow-main-src]][github-workflow-main-href]

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![Renovate - Status][renovate-status-src]][renovate-status-href]
[![License][license-src]][license-href]

> Helps to load the fonts and activate them by preloading.

`nuxt-font-loader-strategy` helps loading the fonts and provides a loading strategy based on preloads.

Define yourself which fonts will be unlocked first.  
This gives the best experience in the initial viewport of the website.

**Features:**
- Use preload to prevent font flashs.
- Generates the `@font-face` definitions automatically and includes them in the layout.  
- Increases the **Pagespeed Insight Score** 🎉
- Take the fonts from [Minimize critical request depth](https://web.dev/critical-request-chains/) and load them via `WebWorker`.
- Deactivate fonts at low connection. (Show [Browser-Support](#browser-support))

> ⚠️ Configuration of the fonts must be included only in the [**module settings**](#options).

[📖 **Release Notes**](./CHANGELOG.md)

## Setup

1. Add `nuxt-font-loader-strategy` dependency to your project

```bash
yarn add nuxt-font-loader-strategy # or npm install nuxt-font-loader-strategy
```

2. Add `nuxt-font-loader-strategy` to the `modules` section of `nuxt.config.js`

```js
{
  modules: [

    ['nuxt-font-loader-strategy', { 
        ignoreLighthouse: true,
        ignoredEffectiveTypes: ['2g', 'slow-2g'],
        fonts: [
          // Font
          {
            fileExtensions: ['woff2', 'woff'],
            fontFamily: 'Font A',
            fontFaces: [
              // Font-Face
              {
                preload: true,
                localSrc: ['Font A', 'FontA-Regular'],
                src: '@/assets/fonts/font-a-regular',
                fontWeight: 400,
                fontStyle: 'normal'
              },
              // Font-Face
              {
                localSrc: ['Font A Light', 'FontA-Light'],
                src: '@/assets/fonts/font-a-300',
                fontWeight: 300,
                fontStyle: 'normal'
              },
              // Font-Face
              {
                localSrc: ['Font A Light Italic', 'FontA-LightItalic'],
                src: '@/assets/fonts/font-a-300Italic',
                fontWeight: 300,
                fontStyle: 'Italic'
              }
            ]
          },
          // Font
          {
            fileExtensions: ['woff2', 'woff'],
            fontFamily: 'Font B',
            fontFaces: [
              // Font-Face
              {
                preload: true,
                src: '@/assets/fonts/font-b-regular',
                fontWeight: 400,
                fontStyle: 'normal'
              },
              // Font-Face
              {
                src: '@/assets/fonts/font-b-700',
                fontWeight: 700,
                fontStyle: 'normal'
              }
            ]
          }
        ]
    }]

  ]
}
```
## Options

| Property                | Type       | Description                                                                                                               | Default                     |
| ----------------------- | ---------- | ------------------------------------------------------------------------------------------------------------------------- | --------------------------- |
| `useWorker`             | `Boolean`  | If set, the non-preloads (prefetches) are loaded via WebWorker.                                                           | `false`                     |
| `ignoreLighthouse`      | `Boolean`  | If set, the non-preloads (prefetches) in Lighthouse are disabled (ignored).                                               | `false`                     |
| `classPattern`          | `Boolean`  | Font css class pattern.                                                                                                   | `[family]_[weight]_[style]` |
| `importPathResolve`     | `Function` | Path resolve for font `src: url(fontPath)`                                                                                | Replace `@/` to `~`         |
| `requirePathResolve`    | `Function` | Path resolve for `require(fontPath)`                                                                                      | no changes                  |
| `ignoredEffectiveTypes` | `Array`    | List of excluded connection types.                                                                                        | `[]`                        |
| `fonts`                 | `Array`    | List of included fonts.                                                                                                   | `[]`                        |
| `unlockDelay`           | `Number`   | Delay in milliseconds for unlock prefetched fonts.                                                                        | `0`                         |
| `prefetchCount`         | `Number`   | Defines how many fonts are prefetched at the same time. <br>**Important:** Lower than zero, everything is loaded at once. | `2`                         |



### Maximum expression `classPattern`

`[family]_[variant]_[featureSettings]_[stretch]_[weight]_[style]`

### WebWorker `useWorker`

Look for compactability at https://github.com/webpack-contrib/worker-loader.

WebWorker is executed with the setting `inline` to reduce the script loads.

### Font

| Property         | Type     | Description      | Default             |
| ---------------- | -------- | ---------------- | ------------------- |
| `fileExtensions` | `Array`  | Font-Family Name | `['woff2', 'woff']` |
| `fontFamily`     | `String` | Font-Family Name | `['2g', 'slow-2g']` |
| `fontFaces`      | `Array`  | Font-Faces       | `[]`                |

### Font-Face

| Property              | Type      | Description                                  | Default    |
| --------------------- | --------- | -------------------------------------------- | ---------- |
| `preload`             | `Boolean` | Specifies whether font is loaded as preload. | `false`    |
| `local`               | `Array`   | List of local font names (System, etc.).     | `[]`       |
| `src`                 | `Array`   | File Path without extension.                 | `null`     |
| `fontVariant`         | `String`  | CSS-Prop. `font-variant`                     | `'normal'` |
| `fontFeatureSettings` | `String`  | CSS-Prop. `font-feature-settings`            | `'normal'` |
| `fontStretch`         | `String`  | CSS-Prop. `font-stretch`                     | `'normal'` |
| `fontWeight`          | `Number`  | CSS-Prop. `font-weight`                      | `'normal'` |
| `fontStyle`           | `String`  | CSS-Prop. `font-style`                       | `'normal'` |
| `fontDisplay`         | `String`  | CSS-Prop. `font-display`                     | `'swap'`   |

>⚠️  **The first `fileExtensions` entry is used as `preload`.**

## Usage

On the HTML tag a class is set for each font file. This class then activates the set styles in the CSS.

The name of the font is specified in `SnakeCase`. (Example: `Open Sans` -> `open_sans`)

**It is recommended to normalize the used tags.**  

**Example:** `h1` has `font-weight: bold` as standard.


```css
p {
  font-family: sans-serif;
}

html.font_open_sans p {
  font-family: 'Roboto', sans-serif;
}
```

For additional FontFaces, classes switch with the options `weight` and `style`.

```CSS
p {
  font-family: sans-serif;
}

html.font_roboto_400_normal p.bold {
  font-family: 'Roboto', sans-serif;
  font-style: normal;
  font-weight: 400;
}

p.bold {
  font-family: sans-serif;
  font-style: normal;
  font-weight: 700;
}

html.font_roboto_700_normal p.bold {
  font-family: 'Roboto', sans-serif;
}

p.light {
  font-family: sans-serif;
  font-style: normal;
  font-weight: 300;
}

html.font_roboto_300_normal p.light {
  font-family: 'Roboto', sans-serif;
}

p.italic {
  font-family: sans-serif;
  font-style: italic;
  font-weight: 400;
}

html.font_roboto_400_italic p.italic {
  font-family: 'Roboto', sans-serif;
}
```


## Browser Performance


![alt text][performance-image]

## Preview

- [Preview](https://grabarzundpartner.github.io/nuxt-font-loader-strategy/)
- [Report Client](https://grabarzundpartner.github.io/nuxt-font-loader-strategy/reports/webpack/client.html)
- [Report Modern](https://grabarzundpartner.github.io/nuxt-font-loader-strategy/reports/webpack/modern.html)
- [Report Server](https://grabarzundpartner.github.io/nuxt-font-loader-strategy/reports/webpack/server.html)


## Development

1. Clone this repository
2. Install dependencies using `yarn install` or `npm install`
3. Start development server using `npm run dev`

## Browser-Support

### Preload Fonts

The options `preload` and `prefetch` are required for the link tag.

**Not all browsers support this:**
- [Can I use - link-rel-preload](https://caniuse.com/#feat=link-rel-preload)
- [Can I use - link-rel-prefetch](https://caniuse.com/#feat=link-rel-prefetch)

If not supported, all fonts are activated.

### Deactivate fonts at low connection

Connection speed dependent font loading, requires the support of `navigator.connection.effectiveType`.

[Can I use - effectivetype](https://caniuse.com/#feat=mdn-api_networkinformation_effectivetype)

## License

[MIT License](./LICENSE)

<!-- Badges -->

[grabarz-partner-module-src]: <https://img.shields.io/badge/Grabarz%20&%20Partner-Module-d19700>
[grabarz-partner-href]: <https://grabarzundpartner.de>

[renovate-status-src]: <https://img.shields.io/badge/renovate-enabled-brightgreen>
[renovate-status-href]: <https://renovate.whitesourcesoftware.com/>

[github-workflow-main-src]: <https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/workflows/Main/badge.svg?branch=main>
[github-workflow-main-href]: <https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/actions?query=workflow%3AMain>

[npm-version-src]: https://img.shields.io/npm/v/nuxt-font-loader-strategy/latest.svg?style=flat-square
[npm-version-href]: https://npmjs.com/package/nuxt-font-loader-strategy

[npm-downloads-src]: https://img.shields.io/npm/dt/nuxt-font-loader-strategy.svg?style=flat-square
[npm-downloads-href]: https://npmjs.com/package/nuxt-font-loader-strategy

[codecov-src]: https://img.shields.io/codecov/c/github/GrabarzUndPartner/nuxt-font-loader-strategy/branch/main/graph/badge.svg?style=flat-square
[codecov-href]: https://codecov.io/gh/GrabarzUndPartner/nuxt-font-loader-strategy

[license-src]: https://img.shields.io/npm/l/nuxt-font-loader-strategy.svg?style=flat-square
[license-href]: https://npmjs.com/package/nuxt-font-loader-strategy

[performance-image]: ./example/static/performance.jpg



CHANGELOG.md
# Project Changelog

## [1.1.10](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.9...v1.1.10) (2021-05-26)


### Bug Fixes

* **npm:** force publish ([d940152](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/d94015229072d142d7eaafc6cf7e9d28cddf532b))

## [1.1.9](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.8...v1.1.9) (2021-03-01)


### Bug Fixes

* **sematic-release:** add pinst to npx ([19575ab](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/19575abaea34e81af843ce354265eb24c9b8bd40))

## [1.1.8](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.7...v1.1.8) (2021-03-01)


### Bug Fixes

* **publish:** force npm publish ([8c0dfe9](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/8c0dfe9c30495b861c8607c2df47162ca853d0b9))

## [1.1.7](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.6...v1.1.7) (2021-03-01)


### Bug Fixes

* **)preload:**  get mimetype from format ([e24e71c](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/e24e71c6315d8c90a4284764190b84378136cd70))

## [1.1.6](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.5...v1.1.6) (2021-02-22)


### Bug Fixes

* **path:** fix extension (ttf) to format (truetype) conversion ([03d7fa1](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/03d7fa1ab461a59c5f684b24ba7c1b898dfa0504))

## [1.1.5](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.4...v1.1.5) (2021-01-29)


### Bug Fixes

* **imports:** fix css import for require and absolute source ([2ed40c9](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/2ed40c993157d91752fb1cc12d7c07c8b152c5dc))
* **path:** fix filepath for windows ([cd3b8f0](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/cd3b8f0bb56fe300e547c8efdd8188a0542102f4))

## [1.1.4](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.3...v1.1.4) (2021-01-29)


### Bug Fixes

* **fonts:** fix absolute font path with router base ([52951bf](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/52951bfeb1ffeecaa072f28198bacd09dda7e8b3))

## [1.1.3](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.2...v1.1.3) (2021-01-29)


### Bug Fixes

* **fonts:** fix font file integration for absolut paths ([1354ec0](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/1354ec0041296046e0440683035873b191f3f41e))

## [1.1.2](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.1...v1.1.2) (2020-11-29)


### Bug Fixes

* **array-flat:** polyfill flat ([9a67da9](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/9a67da980cdf8f0c4257365f463b8b6990ed327e))

## [1.1.1](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.1.0...v1.1.1) (2020-06-04)


### Bug Fixes

* **reports:** fix action for reports ([c71ab69](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/c71ab6926ab8ad6b3223679010739875b14224ba))

# [1.1.0](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/compare/v1.0.0...v1.1.0) (2020-04-28)


### Features

* **lighthouse:** add option to ignore prefetches in lighthouse ([5b5eab8](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/5b5eab8bc31c89e1d58a7529ffa5cbd8e59e4e65))
* **lighthouse:** add option to ignore prefetches in lighthouse ([131b17e](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/131b17e4aabd9204df3cdefb9a10fc1704ea2d45))

# 1.0.0 (2020-02-22)


### Bug Fixes

* **file-ext:** added default file extensions ([8f9fba2](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/8f9fba26d81c88fc717e981b7fc5807dd0163274))
* **travis:** fix gh-pages ([4615d37](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/4615d37c010aad3361b126d1dc767f7acb787e9e))
* **travis-eslint:** fix build task and update eslint plugins ([2b76630](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/2b76630e326c6fc2b97d7d85dc6990a115930651))


### Features

* **initial:** Initial Commit ([8677c1b](https://github.com/GrabarzUndPartner/nuxt-font-loader-strategy/commit/8677c1b9022a146734c02684594974c02a4fdaa6))


example/nuxt.config.js
const { resolve } = require('upath')
const isDev = process.env.NODE_ENV === 'development'
const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin

module.exports = {
  dev: isDev,

  modern: isDev ? false : 'client',

  rootDir: resolve(__dirname, '..'),
  buildDir: resolve(__dirname, '.nuxt'),
  srcDir: __dirname,

  render: {
    resourceHints: false
  },

  env: {
    GITHUB_REPO_URL: `https://github.com/${process.env.GITHUB_REPOSITORY}`
  },

  build: {

    babel: {
      presets ({ isServer }) {
        return [
          [
            require.resolve('@nuxt/babel-preset-app-edge'),
            {
              buildTarget: isServer ? 'server' : 'client',
              corejs: { version: 3 }
            }
          ]
        ]
      }
    },

    filenames: {
      app: ({ isDev }) => isDev ? '[name].js' : '[name].[chunkhash].js',
      chunk: ({ isDev }) => isDev ? '[name].js' : '[name].[chunkhash].js'
    },

    postcss: {
      plugins: {
        'postcss-nesting': {}
      }
    },

    extend (config) {
      if (!isDev) {
        config.plugins.push(new BundleAnalyzerPlugin({
          reportFilename: resolve(`.reports/webpack/${config.name}.html`),
          statsFilename: resolve(`.reports/webpack/stats/${config.name}.json`),
          analyzerMode: 'static',
          generateStatsFile: true,
          openAnalyzer: false,
          logLevel: 'info',
          defaultSizes: 'gzip',
          statsOptions: 'normal'
        }))
      }
    }
  },

  generate: {
    dir: getDistPath()
  },

  router: {
    base: getBasePath()
  },

  modules: [
    [
      resolve(__dirname, '..'), {
        ignoreLighthouse: true,
        prefetchCount: 2,
        fonts: [
          {
            fileExtensions: ['woff2', 'woff'],
            fontFamily: 'Roboto',
            fontFaces: [
              {
                local: ['Roboto', 'Roboto-Regular'],
                src: '@/assets/fonts/roboto-v20-latin/roboto-v20-latin-regular',
                fontWeight: 400,
                fontStyle: 'normal'
              },
              {
                local: ['Roboto Light', 'Roboto-Light'],
                src: '@/assets/fonts/roboto-v20-latin/roboto-v20-latin-300',
                fontWeight: 300,
                fontStyle: 'normal'
              },
              {
                local: ['Roboto'],
                src: '@/assets/fonts/roboto-v20-latin/roboto-v20-latin-italic',
                fontWeight: 400,
                fontStyle: 'italic'
              },
              {
                local: ['Roboto Bold', 'Roboto-Bold'],
                src: '@/assets/fonts/roboto-v20-latin/roboto-v20-latin-700',
                fontWeight: 700,
                fontStyle: 'normal'
              }
            ]
          },
          {
            fileExtensions: ['woff2', 'woff'],
            fontFamily: 'Roboto Slab',
            fontFaces: [
              {
                preload: true,
                src: '/fonts/roboto-slab-v11-latin/roboto-slab-v11-latin-regular',
                fontWeight: 400,
                fontStyle: 'normal'
              },
              {
                preload: true,
                src: '/fonts/roboto-slab-v11-latin/roboto-slab-v11-latin-700',
                fontWeight: 700,
                fontStyle: 'normal'
              }
            ]
          },
          {
            fileExtensions: ['ttf'],
            fontFamily: 'Roboto Mono',
            fontFaces: [
              {
                preload: true,
                src: '@/assets/fonts/roboto-mono/RobotoMono-Regular',
                fontWeight: 400,
                fontStyle: 'normal'
              },
              {
                preload: true,
                src: '@/assets/fonts/roboto-mono/RobotoMono-Bold',
                fontWeight: 700,
                fontStyle: 'normal'
              }
            ]
          }
        ]
      }
    ]
  ]
}

function getBasePath () {
  return process.env.npm_config_base || process.env.BASE_PATH || '/'
}

function getDistPath () {
  return process.env.npm_config_dist || process.env.DIST_PATH || 'dist'
}
